File: HomelabSwift/Homelab/Utilities/Logger.swift
import Foundation
import os

/// A centralized structured logger for the Homelab application.
public struct AppLogger: Sendable {
    public static let shared = AppLogger()
    
    private let logger = Logger(subsystem: Bundle.main.bundleIdentifier ?? "com.homelab.app", category: "General")
    
    public func debug(_ message: String) {
        logger.debug("\(message, privacy: .public)")
    }
    
    public func info(_ message: String) {
        logger.info("\(message, privacy: .public)")
    }
    
    public func error(_ message: String) {
        logger.error("\(message, privacy: .public)")
    }
    
    public func error(_ error: Error) {
        logger.error("Error: \(error.localizedDescription, privacy: .public)")
    }

    /// Logs state transitions for ViewModels using LoadableState
    public func stateTransition<T>(service: String, state: LoadableState<T>) {
        let stateString: String
        switch state {
        case .idle: stateString = "Idle"
        case .loading: stateString = "Loading"
        case .loaded: stateString = "Loaded"
        case .error(let err): stateString = "Error (\(err.errorDescription ?? "unknown"))"
        case .offline: stateString = "Offline"
        }
        logger.debug("[\(service, privacy: .public)] State Transition -> \(stateString, privacy: .public)")
    }
}

File: HomelabAndroid/app/src/main/java/com/homelab/app/util/Logger.kt
package com.homelab.app.util

import android.util.Log

/**
 * Centralized logging utility for the Homelab Android application.
 */
object Logger {
    fun d(tag: String, message: String) {
        Log.d(tag, message)
    }

    fun i(tag: String, message: String) {
        Log.i(tag, message)
    }

    fun w(tag: String, message: String) {
        Log.w(tag, message)
    }

    fun e(tag: String, message: String, throwable: Throwable? = null) {
        Log.e(tag, message, throwable)
    }

    fun stateTransition(tag: String, stateName: String, state: UiState<*>) {
        val stateString = when (state) {
            is UiState.Idle -> "Idle"
            is UiState.Loading -> "Loading"
            is UiState.Success -> "Success"
            is UiState.Error -> "Error(${state.message})"
            is UiState.Offline -> "Offline"
        }
        Log.d(tag, "State Transition -> $stateName: $stateString")
    }
}
